File: Different problems/copilot-BD.py
# def point_in_polygon(vertices, point):
#     n = len(vertices)
#     crossings = 0

#     for i in range(n):
#         x1, y1 = vertices[i]
#         x2, y2 = vertices[(i + 1) % n]

#         if (y1 > point[1]) != (y2 > point[1]):
#             x_intersection = x1 + (x2 - x1) * (point[1] - y1) / (y2 - y1)
#             if point[0] < x_intersection:
#                 crossings += 1

#     return "YES" if crossings % 2 == 1 else "NO"


# point = input().split()

# n = int(point[0])
# point = tuple(map(int, point[1:]))

# vertices = []
# for _ in range(n):
#     vertices.append(tuple(map(int, input().split())))

# print(point_in_polygon(vertices, point))


def is_inside_polygon(vertices, point):
    n = len(vertices)
    inside = False

    for i in range(n):
        x1, y1 = vertices[i]
        x2, y2 = vertices[(i + 1) % n]

        if (y1 > point[1]) != (y2 > point[1]):
            x_intersection = x1 + (x2 - x1) * (point[1] - y1) / (y2 - y1)
            if point[0] < x_intersection:
                inside = not inside

    return inside


point = input().split()

n = int(point[0])
point = tuple(map(int, point[1:]))

vertices = []
for _ in range(n):
    vertices.append(tuple(map(int, input().split())))

print("YES") if is_inside_polygon(vertices, point) else print("NO")

File: Different problems/copilot-BD.cpp
#include <iostream>
#include <vector>
using namespace std;

struct Point
{
    int x, y;
};

bool isInsidePolygon(const vector<Point> &polygon, const Point &p)
{
    int n = polygon.size();
    int cnt = 0; // Количество пересечений луча

    for (int i = 0; i < n; ++i)
    {
        const Point &a = polygon[i];
        const Point &b = polygon[(i + 1) % n];

        if ((a.y <= p.y && b.y > p.y) || (a.y > p.y && b.y <= p.y))
        {
            // Проверяем, пересекает ли ребро луч
            double xIntersect = (a.x * (b.y - p.y) + b.x * (p.y - a.y)) / (double)(b.y - a.y);
            if (p.x < xIntersect)
                ++cnt;
        }
    }

    return cnt % 2 == 1; // Нечётное количество пересечений
}

int main()
{
    int n;
    Point p;
    cin >> n >> p.x >> p.y;

    vector<Point> polygon(n);
    for (int i = 0; i < n; ++i)
    {
        cin >> polygon[i].x >> polygon[i].y;
    }

    if (isInsidePolygon(polygon, p))
    {
        cout << "YES\n";
    }
    else
    {
        cout << "NO\n";
    }

    return 0;
}
